src/callbacks_stage/subcell_bounds_check.jl

```diff
@@ -0,0 +1,124 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+#! format: noindent
+
+"""
+    BoundsCheckCallback(; output_directory="out", save_errors=false, interval=1)
+
+Bounds checking routine for [`SubcellLimiterIDP`](@ref). Applied as a stage callback for SSPRK
+methods. If `save_errors` is `true`, the resulting deviations are saved in
+`output_directory/deviations.txt` for every `interval` time steps.
+
+!!! note
+    For `SubcellLimiterIDP`, the solution is corrected in the a posteriori correction stage
+    [`SubcellLimiterIDPCorrection`](@ref). So, to check the final solution, this bounds check
+    callback must be called after the correction stage.
+"""
+struct BoundsCheckCallback
+    output_directory::String
+    save_errors::Bool
+    interval::Int
+end
+
+function BoundsCheckCallback(; output_directory = "out", save_errors = false,
+                             interval = 1)
+    BoundsCheckCallback(output_directory, save_errors, interval)
+end
+
+function (callback::BoundsCheckCallback)(u_ode, integrator, stage)
+    mesh, equations, solver, cache = mesh_equations_solver_cache(integrator.p)
+    (; t, iter, alg) = integrator
+    u = wrap_array(u_ode, mesh, equations, solver, cache)
+
+    save_errors_ = callback.save_errors && (callback.interval > 0) &&
+                   (stage == length(alg.c)) &&
+                   (iter % callback.interval == 0 || integrator.finalstep)
+    @trixi_timeit timer() "check_bounds" check_bounds(u, mesh, equations, solver, cache,
+                                                      solver.volume_integral, t,
+                                                      iter + 1,
+                                                      callback.output_directory,
+                                                      save_errors_)
+end
+
+function check_bounds(u, mesh, equations, solver, cache,
+                      volume_integral::AbstractVolumeIntegral, t, iter,
+                      output_directory, save_errors)
+    return nothing
+end
+
+function check_bounds(u, mesh, equations, solver, cache,
+                      volume_integral::VolumeIntegralSubcellLimiting, t, iter,
+                      output_directory, save_errors)
+    check_bounds(u, mesh, equations, solver, cache, volume_integral.limiter, t, iter,
+                 output_directory, save_errors)
+end
+
+function init_callback(callback, semi)
+    init_callback(callback, semi, semi.solver.volume_integral)
+end
+
+init_callback(callback, semi, volume_integral::AbstractVolumeIntegral) = nothing
```

Comment:
These seem to me like very generic implementation, same as `finalize_callback`. Are these functions you introduced or are they prescribed by OrdinaryDiffEq?

If these are Trixi-specific functions, I think they should rather be put in a more central location than this very specific stage callback.

cc @ranocha

Reply:
I agree with you that they look like a generic implementation.
On the other hand, there are only 2 stage callbacks that use this specific form with `init/finalize_callback` anyway. The other one (`SubcellLimiterIDPCorrection`) doesn't do anything at all for these two routines and has its own implementation with
```
init_callback(limiter!::SubcellLimiterIDPCorrection, semi) = nothing
finalize_callback(limiter!::SubcellLimiterIDPCorrection, semi) = nothing
```

Maybe it would be useful to just dispatch those 2 function for the specific stage callback here for now (with the callback name). And if there will be another callback like this in the future (there is none planed for subcell limiting), we can move the functions to a more generic location. What do you think about that?

Reply:
So you mean, add an `init/finalize_callback(limiter!::BoundsCheckLImiter, ...)` instead of the generic versions? That sounds much better imho

Reply:
Yes, exactly:+1: Then it is clear that this function is only for this callback.

And while we're at it, it's probably also more useful to write `init/finalize_callback(callback::BoundsCheckCallback, ...)` instead of `limiter`. And I will change this for `SubcellLimiterIDPCorrection` as well, since this is also a callback.